### PR TITLE
tests: working example Terraform

### PR DESCRIPTION
**Pull request by paulschwarzenberger**
## what
- working and tested example Terraform

## why
- example for users of the module
- basis for automated tests (to follow)

### CODE REVIEW

File: examples/complete/README.md [hunk 1]
```diff
@@ -1 +1,12 @@
 # Example deployment
+
+## Local testing
+* replace the Terraform state S3 bucket fields in the command below as appropriate
+* for local testing, duplicate `terraform.tfvars.example`, rename without the `.example` suffix
+* enter details appropriate to your organization and save
+
+```bash
+terraform init -backend-config=bucket=TERRAFORM_STATE_BUCKET -backend-config=key=TERRAFORM_STATE_KEY -backend-config=region=TERRAFORM_STATE_REGION
+terraform workspace new dev
```
Comment (nitrocode) [inline examples/complete/README.md]: Can we remove the need for a workspace with local testing?

```suggestion
```

File: examples/complete/main.tf [hunk 2]
```diff
@@ -1,8 +1,25 @@
 module "domain_protect" {
-  source = "../../"
+  #source = "../../"
+  source  = "domain-protect/domain-protect/aws"
```
Comment (nitrocode) [inline examples/complete/main.tf:3]: The hope with the `examples/complete` is to run integration tests for each PR so it should continue to use the relative path than the published path.

Maybe we can do this instead?

```suggestion
  source = "../../"

  # source  = "domain-protect/domain-protect/aws"
```

File: examples/complete/main.tf [hunk 3]
```diff
@@ -1,8 +1,25 @@
...
+  version = "0.5.1"
```
Comment (nitrocode) [inline examples/complete/main.tf]: The hope with the `examples/complete` is to run integration tests for each PR so it should continue to use the relative path than the published path.

Maybe we can do this instead?

```suggestion
  source = "../../"

  # source  = "domain-protect/domain-protect/aws"
  # version = "0.5.1"
```

File: examples/complete/main.tf [hunk 4]
```diff
@@ -1,8 +1,25 @@
...
+  ip_address          = true
+  ip_scan_schedule    = "10 minutes"
+  ip_time_limit       = "0.1" #6 minutes
+  org_primary_account = var.org_primary_account
+  rcu                 = 1
+  slack_channels      = terraform.workspace == "prd" ? ["devsecops"] : ["devsecops-dev"]
+  scan_schedule       = "10 minutes"
+  slack_webhook_type  = "app"
+  slack_webhook_urls  = var.slack_webhook_urls
+  takeover            = terraform.workspace == "prd" ? true : false
```
Comment (nitrocode) [inline examples/complete/main.tf]: One of the goals was to remove the workspace and environment specific conditions here.

What benefit is there to adding these back in?

File: examples/complete/terraform.tfvars.example [hunk 5]
```diff
@@ -0,0 +1,6 @@
+org_primary_account = "012345678901"
+external_id         = "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
+cf_api_key          = "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
+hackerone_api_token = "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
+
+slack_webhook_urls = ["https://hooks.slack.com/services/XXXXXXXXXX/XXXXXXXXXXXXXXXXX"] # dev
```
Comment (nitrocode) [inline examples/complete/terraform.tfvars.example]: Are these tests planning to test slack, hackerone, cloudflare, etc ? If not, I'd pull these values.

File: examples/complete/main.tf (hunk 3 above)
Comment (paulschwarzenberger) [reply examples/complete/main.tf]: thanks, updated

File: examples/complete/main.tf (hunk 4 above)
Comment (paulschwarzenberger) [reply examples/complete/main.tf]: removed Terraform workspace and conditions

File: examples/complete/terraform.tfvars.example (hunk 5 above)
Comment (paulschwarzenberger) [reply examples/complete/terraform.tfvars.example]: I think it's important to keep these in - initially the test will be a Terraform plan / apply - and it's a more comprehensive test if it includes some options such as CloudFlare and HackerOne. Slack notifications are in any case a core functionality of the system.

I'm also planning for now to deploy to my existing test environment, and I don't want to lose the CloudFlare / HackerOne / Slack functionality!

File: examples/dev/README.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# Example deployment
+
+## Local testing
+* replace the Terraform state S3 bucket fields in the command below as appropriate
+* for local testing, duplicate `terraform.tfvars.example`, rename without the `.example` suffix
+* enter details appropriate to your organization and save
+
+```bash
+terraform init -backend-config=bucket=TERRAFORM_STATE_BUCKET -backend-config=key=TERRAFORM_STATE_KEY -backend-config=region=TERRAFORM_STATE_REGION
+terraform workspace new dev
```
Comment (nitrocode) [inline examples/dev/README.md:14]:
```suggestion
```

File: examples/dev/main.tf [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  hackerone           = "enabled"
+  ip_address          = true
+  ip_time_limit       = 0.1 # 6 minutes
+  org_primary_account = var.org_primary_account
+  rcu                 = 1
+  slack_channels      = ["devsecops-dev"]
+  slack_webhook_type  = "app"
+  slack_webhook_urls  = var.slack_webhook_urls
+  takeover            = false
+  wcu                 = 1
```
Comment (nitrocode) [inline examples/dev/main.tf:20]: Why not set all these in tfvars so each tfvars file can be a separate test?

File: examples/dev/terraform.tfvars.example
Comment (nitrocode) [inline examples/dev/terraform.tfvars.example:1]: I wouldn't do a `examples/<env>` instead, do a `examples/complete/env.tfvars.example` so you can reduce the code

File: examples/dev/main.tf (hunk 7 above)
Comment (paulschwarzenberger) [reply examples/dev/main.tf:20]: please see discussion on Slack

File: examples/dev/terraform.tfvars.example
Comment (paulschwarzenberger) [reply examples/dev/terraform.tfvars.example:1]: please see discussion on Slack

File: examples/dev/README.md (hunk 6 above)
Comment (paulschwarzenberger) [reply examples/dev/README.md:14]: updated to show use of Terraform workspace is optional, and to explain the circumstance in which it's useful - if a single Terraform state bucket is used for multiple environments.

**Review by nitrocode — APPROVED** [review]

**Pull request merged**